### ISSUE TITLE
Multiple `assert_email_sent` calls in one test appears sometimes fails

### ISSUE THREAD

**Issue by yukster** [swoosh/swoosh#488]
I have a test that exercises some code that can send multiple emails, getting those addresses from different sources. My test setup creates a situation where two emails will be sent, one from each source, and I have two calls to `assert_email_sent` to ensure that both cases got their email and it generally passes locally but fails in CI consistently.

I've put a `receive` block in the test as well as debugging output in my code right after the `deliver` call. All as is expected in my debugging output: both emails are definitely getting queued in the test adapter. But the receive block only gets called once. 

It appears the `receive` block messes things up as well, cuz the test fails locally with it in there. I was just curious if I'd receive two messages and I don't. I see the receive call for the the email in the first assert and the test fails on the second. Again, I am sure that `deliver` actually gets called twice. But it's weird that the test can pass if there aren't actually two messages being received.

Maybe there is some case where `:"$callers"` doesn't do what we expect?:
https://github.com/swoosh/swoosh/blob/5a278c7959d9b704530ec3fe6b6df53fe3d47ca5/lib/swoosh/adapters/test.ex#L32

Or maybe this issue is actually with assert_receive?

Ultimately I just stripped the test down to one assertion, but it seems sad to only be able to call it once per test!

**Comment by princemaple** [comment]
It'd be very helpful if you could post your test code.

Receive block does not automatically repeat. It simply receives one message. By receiving it, it consumes it in the message queue, so `assert_receive` in `assert_email_sent` will not be able to receive it.

**Comment by yukster** [comment]
Wow, thanks for the fast reply! I'm not sure what test code I could post because it's just creating a few factory records, then calling the function that should result in the emails getting sent, then two calls to `assert_email_sent`. Not gonna mean much to ya. And it would be pretty onerous to try to copy and paste all the places in the code path that are required to follow the flow. 

But yeah, I get ya that putting in a receive block messes things up... was just an attempt to debug. 

I think this all boils down to a fairly simple question/issue: I'm sure my code is pushing two emails to the test adapter and the test with two asserts passes locally but fails in CI. I'm wondering if anyone has run into issues with doing multiple asserts and if there is a bug in the test adapter  because it's possible to have a CI environment not get all the messages?

I was toying with the idea of writing an alternate test_assertions module that uses Swoosh.Adapters.Local.Storage.Memory instead. Maybe that would be more robust? Or maybe starting/stopping it or at least flushing its queue after every test would be a pain?

**Comment by princemaple** [comment]
> I think this all boils down to a fairly simple question/issue: I'm sure my code is pushing two emails to the test adapter and the test with two asserts passes locally but fails in CI. I'm wondering if anyone has run into issues with doing multiple asserts and if there is a bug in the test adapter because it's possible to have a CI environment not get all the messages?

This really shouldn't be an issue. Message sending is very fast, even on CI. Receive can cop with different orders of arrival of messages as well.



> I was toying with the idea of writing an alternate test_assertions module that uses Swoosh.Adapters.Local.Storage.Memory instead. Maybe that would be more robust? Or maybe starting/stopping it or at least flushing its queue after every test would be a pain?

This could be a good idea. Use with start_supervised, etc, can be very powerful.

**Comment by yukster** [comment]
Ok, I will try to look into trying to use the memory storage for tests. It would be really nice to be able to assert the number of emails sent, and the current assertions can't do that.

And maybe at some point I can dig through our CI set u and try to figure out why multiple assert_email_sent calls won't work... but I promise you they consistently do not.

**Comment by princemaple** [comment]
Thanks.

I believe you. I just can't do anything if I don't see the code.

**Comment by zorn** [comment]
One behavior that might be related that I did not expect, if you send two emails and then call `assert_email_sent` with like a `subject` check, it will only assert the first sent email subject line. If you try to assert the second email subject line it will fail. More so, once you assert the first email subject line if you try to do the assert again it fails.

Best I can tell, the test adapter will only store one email at a time and it's the first email that was sent. Once you use `assert_email_sent` it clears that memory.

Still learning the library so I might be wrong here. Would welcome more options for email testing. I myself have behaviors that kick off two emails (one to admin one to user) and I need to assert both.

Anyways, thought I'd share my hopefully correct observations.

**Comment by yukster** [comment]
@zorn: interesting! The two asserts in this test were for the same email copy but to different addresses. I did have the subject in the assert. I'm pretty sure, however, that while trying all sorts of things to fix the problem, I tried it with just the `email: foo` arg on both. And then there's the fact that the test passed locally but failed in CI. That's super-weird.

**Comment by princemaple** [comment]
Test assertion when used in a way like `assert_email_sent subject: "Hello, Avengers!"`, it actually receives any arbitrary email then tries to match on the properties. That's probably the source of confusion in many cases. The reason is, AFAIK, you cannot dynamically build up a pattern.

It indeed seems now that a new test adapter that is a variation of the local adapter may be more suitable for better testing experience.

**Comment by josevalim** [comment]
The current testing functionality may have race conditions if you are sending emails asynchronously, as it does not wait in the inbox. A configurable timeout could be given but I wonder if we should introduce a new set of testing helpers that are macro based. The idea is:

```
assert_email_sent %{subject: "You got a message"}
```

Is a macro that will expand to:

```
assert_received {:email, %{subject: "You got a message"}}
```

This means we can also have a:

```
assert_email_sending %{subject: "You got a message"}, 1000
```

Which will compile to:

```
assert_receive {:email, %{subject: "You got a message"}}
```

This means you can match on emails in whatever order and with arbitrary timeouts. We can also make it so `assert_email_sent/sending` returns the email, so you can do further assertions:

```
email = assert_email_sending %{subject: "You got a message"}, 1000
assert email.to == {"", "foo@example.com"}
```
However, this has a couple downsides:

1. Since it is pattern matching, you need to match on the inner recipient format of emails:

```
assert_receive {:email, %{to: {_, "foo@example.com"}}
```

2. Since it is pattern matching, you cannot have interpolations. For example, to match on an email given by the system:

```
email = user.email
assert_email_sent %{to: {_, ^email}}
```

or

```
assert_email_sent %{to: {_, email}} when email == user.email
```

This is exactly the confusion people have on #797 but perhaps it can all be addressed in the docs. Phoenix uses this same technique for testing channels and push messages from LiveView. A function called `flush_emails()` that returns all emails could also be kept if someone prefers the current style of assertions:

```
assert flush_emails() |> Enum.any?(& &1.subject =~ "hello")
```

I will play with these for Livebook Teams and if it works out I will paste some code in here later. :)

**Comment by AJReade** [comment]
Hey guys, 

Is there a reason this hasn't moved into the library?

**Comment by princemaple** [comment]
@AJReade a few reasons:

1. it's a breaking change
2. there are a few different versions floating around
  - the "legacy" macro-based test assertions, with blind `receive`
  - the relatively new `X.TestAssertion` pure function based, with blind `receive`, cleans up macros and is backwards compatible but does not solve race condition issue
  - José's `receive` version with pattern matching, solves race condition but is breaking change
3. There are more requirements for integration tests, which all options in (2) don't satisfy

**Comment by AJReade** [comment]
@princemaple Thanks for letting me know. Is there a way to do assert_email_sending in the current library? This is the macro I have personal made use of and the rest I reach directly in to the lib for.

**Comment by princemaple** [comment]
For this specific purpose I recommend copying what José posted and simply use it everywhere in your tests.

**Comment by AJReade** [comment]
I see that's what I've done already. If there's no way to do this or something comparable in the library, are there any plans for add additional testing utilities to address the gap?

**Comment by princemaple** [comment]
Honestly I haven't got the time to think thoroughly about this. I could simply bring what José has here into another separate module but we will have 3 different ones then. I'd really like to settle on one universal solution.

**Comment by AJReade** [comment]
Yes I agree a universal module that is fully capable would be ideal, thanks for sharing your thoughts anyway 🙌🏼
